Project1

GaxysGalactic opened the pull request:
ShapeBuilder Project (std::array and std::vector)

GaxysGalactic committed:
Part 1 - Classes

+ Added Shape (abstract) class
+ Added Circle class
+ Added Square class

GaxysGalactic committed:
Part 1 Keyword fixes

o Fixes on some keyword usage for Part 1

GaxysGalactic committed:
Part 2 - Main Menu

+ Added ShapeBuilder to organize App
+ Added Comment separators
+ Added MainMenu and appropriate methods
+ Added Utility Functions

GaxysGalactic committed:
Part 3 - Built In Array

+ Added Built In Array functionality
o Fixed keywords again on overrides

GaxysGalactic committed:
Part 4 - StaticArray Header

+ Added header for TStaticArray

GaxysGalactic committed:
Part 5 - StaticArray Methods I

+ Implemented functionality for methods in the StaticArray template
+ Added a delete function for heap stored arrays

GaxysGalactic committed:
Part 6 - StaticArray Methods II

+ Implemented additional methods for StaticArray

GaxysGalactic committed:
Part 7

+ Implemented functionality for Static Array
o Fixed Swap function in template

GaxysGalactic committed:
Part 7 - TDynamicArray Basics

+ Created the template class file for TDynamicArray
+ Added constructor, destructor, and properties

GaxysGalactic committed:
Part 9 - Access

+ Added methods that implement accessing functionality into the Dynamic Array

GaxysGalactic committed:
Part 10 - Capacity

+ Added methods that affect capacity and resize the array if prompted

GaxysGalactic committed:
Part 11 - Modifiers

+ Added modifiers to the Dynamic Array
o Fixed missing argument in Static Array method

GaxysGalactic committed:
Part 13 - Custom methods

+ Added custom methods to the Dynamic Array class.

GaxysGalactic committed:
Bug Fixes and Setup

+ Added methods and organized them to prepare for Dynamic Array Implementation
o Fixed typos in certain parts of DynamicArray

GaxysGalactic committed:
Part 14 - Finishing Touches

+ Added implementation of TDynamicArray in the ShapeBuilder class
o Fixed copymaking and reference issues on some methods of the templates
o Fixed const issues

GaxysGalactic committed:
Bug Fix and Renaming

o Fixed bug with operator signature
o Renamed some functions to have better use of overloading.

simonkytr commented on Projects/Shapes/Shapes/Shapes/Circle.cpp:
```diff
@@ -0,0 +1,27 @@
+//-------------------------------------------------------------------------------------------
+float FCircle::GetArea() const
+{
+    return 3.14159265f * Radius * Radius;
```
you should use the function `pow(Radius, 2)`

simonkytr commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,339 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Reserve(const int NewCapacity)
```
I like this method to be a boolean function but, I think when the new capacity is smaller than capacity, the method shouldn't do anything, so I'm not sure that method should return an boolean value, it is just a comment, maybe I am wrong

JDAlzate replied:
![AgreedGIF](https://user-images.githubusercontent.com/36243539/191055707-2ab91e93-32cc-4a9b-b5a3-b116f9aad5ac.gif)

simonkytr commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,339 @@
+
+	TDynamicArray(const int InCapacity)
+	{
+		Capacity = InCapacity;
```
you're introducing a capacity to a dynamic array, but it supposes to be created without to necessity to be initialized with some capacity size, therefore when you initialize the variable capacity, it should have a value

GaxysGalactic replied:
Capacity and Size are by default 0, initialized in the member variables above, according to our coding standards. The two constructors available for TDynamicArray are:
- Default Constructor -> This one is empty, as it would be a waste if it initialized anything by default. If I were to make an Array of N TDynamicArrays, it would instantly call N constructors. If the default constructor reserves something by default, the time and space complexity would be affected.
- Capacity Constructor -> This one can be called if you already have a good estimate of the capacity your array is going to need. This allows for optimization in space complexity. If you are unsure of the capacity it should initially have, it would be better to call the Default Constructor.

The downside to this structure of constructors is the additional if clause check in the Push function. However, the time complexity loss there is negligible compared to that saved by the default constructor change.

simonkytr commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,339 @@
+	//-------------------------------------------------------------------------------------------
+	void PopBack()
+	{
+		Size--;
```
what happens if your array size is 0?

simonkytr commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,339 @@
+		if (Index == Size)
+		{
+			PopBack();
+			Size--;
```
the method `PopBack()` and `Size--` isn't doing the same?

simonkytr commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,339 @@
+				Array[i] = Item;
+			}
+		}
+	}
```
all items between size and New size going to contain the same item, why?

GaxysGalactic replied:
This one is a function overload based on the std::vector function. On std::vector, there is an overload so that you are able to not only resize, but fill any missing slots with an item of your choice. This method is part of the "creative" methods for part 13

JDAlzate replied:
![ILikeThisAlanShaneLewisGIF](https://user-images.githubusercontent.com/36243539/191057283-66777f80-60dc-44ad-8ce2-5a7a9b42831f.gif)

JDAlzate replied:
You probably should've put this in the Custom Methods category though

simonkytr commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,339 @@
+	//-------------------------------------------------------------------------------------------
+	void Append(const TDynamicArray<T>& OtherArray)
+	{
+		int NewSize = Size + OtherArray.GetSize();
```
you could do this operation straightly in Reserve;

simonkytr commented on Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp:
```diff
@@ -0,0 +1,369 @@
+	return Option;
+}
+
+
```
the functions print or create shape are doing the same, you could merge them in just one function, you should create the `for` inside the ask" type" array method and you don't need to overlap the function

simonkytr reviewed (changes requested):
Niiice I like this work so much ... some functions in the shape builder are doing exactly the same, could you optimize these code lines?, but with the rest, I learned so much with your work, good job

GaxysGalactic committed:
Code Review Fixes

o Fixed the error in PopBack() and Erase()
o Small optimization in Circle.cpp
o Small optimization in Append()

JDAlzate commented on Projects/Shapes/Shapes/Shapes/Square.cpp:
```diff
@@ -0,0 +1,27 @@
+//-------------------------------------------------------------------------------------------
+FSquare::FSquare()
+{
+    Side = 0;
```
Missing .f here.

JDAlzate commented on Projects/Shapes/Shapes/Shapes/Circle.cpp:
```diff
@@ -0,0 +1,28 @@
+//-------------------------------------------------------------------------------------------
+FCircle::FCircle()
+{
+    Radius = 0;
```
Missing .f

JDAlzate commented on Projects/Shapes/Shapes/Shapes/Circle.cpp:
```diff
@@ -0,0 +1,28 @@
+//-------------------------------------------------------------------------------------------
+float FCircle::GetPerimeter() const
+{
+    return 2 * 3.14159265f * Radius;
```
Missing .f

JDAlzate commented on Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp:
```diff
@@ -0,0 +1,369 @@
+		delete Array[i];
+		Array[i] = nullptr;
+	}
+	delete Array;
```
Careful here! Remember that to delete arrays you need to use `delete[]`, not just `delete`. It may work but might just as well not work, this is undefined behavior.

JDAlzate commented on Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp:
```diff
@@ -0,0 +1,369 @@
+}
+
+//-------------------------------------------------------------------------------------------
+TStaticArray<FShape*, 5> FShapeBuilder::FillStaticArray() const
```
This is fine and working perfectly. However, when we want to return something like a container, it may be better to use out parameters.
```
void FillStaticArray(TStaticArray<FShape*, 5>& OutArray) const;
```
because when you're creating an object for the return type and then, when you return, calling a copy assignment or copy constructor to assign the returned variable to your actual variable outside the function.

By using out parameters, you remove one of those copy operations.

Compilers are pretty smart these days and may choose to optimize this. However, you can't know for sure, so it may be better to not assume that the compiler will optimize the return type and use an out parameter instead.

We can discuss this in person if maybe the explanation is not too clear :)

GaxysGalactic replied:
Yes, I saw that some extra copies were created, but at first I tried with references, though, being an object created in the method itself, that didn't work as planned. I forgot about OutParameters and then just defaulted to returning a copy instead 😢

JDAlzate commented on Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp:
```diff
@@ -0,0 +1,369 @@
+}
+
+//-------------------------------------------------------------------------------------------
+void FShapeBuilder::PrintArray(TStaticArray<FShape*, 5>& Array) const
```
This array is not changing and should be a const reference 👀

JDAlzate commented on Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp:
```diff
@@ -0,0 +1,369 @@
+	TDynamicArray<FShape*> Shapes;
+
+
+	int Continue = 1;
```
An int called continue doesn't tell me anything. Maybe this could be a bool? And so you could assign to the bool inside the loop.

JDAlzate commented on Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp:
```diff
@@ -0,0 +1,369 @@
+}
+
+//-------------------------------------------------------------------------------------------
+void FShapeBuilder::PrintArray(TDynamicArray<FShape*>& Array) const
```
This array is not changing and should be a const reference 👀

JDAlzate commented on Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp:
```diff
@@ -0,0 +1,369 @@
+}
+
+//-------------------------------------------------------------------------------------------
+TDynamicArray<FShape*> FShapeBuilder::FillDynamicArray() const
```
Same comment here about maybe using an out parameter

JDAlzate commented on Projects/Shapes/Shapes/Shapes/Main.cpp:
```diff
@@ -0,0 +1,8 @@
+
+int main()
+{
+	FShapeBuilder ShapeBuilderApp = FShapeBuilder();
```
Remember that the `= FShapeBuilder()` is not necessary. You're calling a copy constructor and passing in a default constructed shape.

The compiler will most likely optimize it, but still, it's worth knowing about it.

JDAlzate commented on Projects/Shapes/Shapes/Shapes/StaticArray.h:
```diff
@@ -0,0 +1,92 @@
+class TStaticArray
+{
+private:
+	int Size = N;
```
It may be worth it to have this variable be const or even const static since it will never change

JDAlzate commented on Projects/Shapes/Shapes/Shapes/StaticArray.h:
```diff
@@ -0,0 +1,92 @@
+	};
+
+	//-------------------------------------------------------------------------------------------
+	void Fill(T& Filler)
```
Is this Filler changing?
![image](https://user-images.githubusercontent.com/36243539/191055254-13433a3d-8c8c-400e-a45d-e440743fc494.png)

GaxysGalactic replied:
![image](https://user-images.githubusercontent.com/44332093/191064879-cc052f16-9add-4ef6-a7d6-0288542fa971.png)

In other news I'm pretty sure I improved on the const parameters for the next project. On the primitive data types I remember to place const but somehow for the ones that matter the most I forget. I think it's related to the color coding (const and primitive data types are same color x3 )

JDAlzate commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,341 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool ShrinkToFit()
```
I have the same comment here as the one that Diego had before

JDAlzate commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,341 @@
+			return true;
+		}
+
+		if (Size == Capacity)
```
This one is dangerous considering that your initial capacity and initial size are both 0 if not specified otherwise.

JDAlzate commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,341 @@
+			return false;
+		}
+
+		if (Index == Size)
```
Careful here. This works on insert because you want to insert in an index immediately after the last one, but how can you erase the last index by asking to delete the last + 1?

JDAlzate commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,341 @@
+	//											CUSTOM METHODS
+	//-------------------------------------------------------------------------------------------
+
+	void RefactorCapacityTo(const int NewCapacity)
```
Hmmm looks good but unnecessarily complicated. A resize and a shrink-to-fit would do the trick.

GaxysGalactic replied:
I can see how it would work, would it be something similar to this?

Say you have a vector of items called Inventory, which had its initial capacity as 30 since that's the maximum capacity at the start of the game. After an unfortunate accident, the player now has their inventory reduced to a maximum of 10. 

```
//Game Start
TDynamicArray<Item> Inventory(30);
...
//Unfortunate Accident
int PreviousSize = Inventory.GetSize();
Inventory.Resize(10);
Inventory.ShrinkToFit();
Inventory.Resize(PreviousSize);
```

I assume that you would have to store the previous size, as you would want the size to be correct after the capacity change. Or is there a better way? The "better way" that I thought was this method, but I could be wrong: I remember when browsing online someone mentioned doing this with a swap, but I wasn't too sure how that changed things.

JDAlzate commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,341 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Fill(const T& Item, const bool FillToCapacity)
```
bools need to start with b

JDAlzate commented on Projects/Shapes/Shapes/Shapes/DynamicArray.h:
```diff
@@ -0,0 +1,341 @@
+	//-------------------------------------------------------------------------------------------
+	void Fill(const T& Item, const bool FillToCapacity)
+	{
+		if (FillToCapacity)
```
Nice, but this is the perfect scenario for a ternary operation. When your code is exactly the same except for the value of one variable, it may be better to simply do something like
```
const int FillLimit = bFillToCapacity ? Capacity : Size;
for (int i = 0; i < FillLimit; ++i)
```

That way you avoid long functions and code duplication

JDAlzate reviewed (commented):
Nice!
